// @antv/graphin/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Behaviors", {
  enumerable: true,
  get: function get() {
    return _behaviors.default;
  }
});
Object.defineProperty(exports, "Components", {
  enumerable: true,
  get: function get() {
    return _components.default;
  }
});
Object.defineProperty(exports, "G6", {
  enumerable: true,
  get: function get() {
    return _g.default;
  }
});
Object.defineProperty(exports, "GraphinContext", {
  enumerable: true,
  get: function get() {
    return _GraphinContext.default;
  }
});
Object.defineProperty(exports, "Utils", {
  enumerable: true,
  get: function get() {
    return _utils.default;
  }
});
exports.registerFontFamily = exports.default = void 0;

var _behaviors = _interopRequireDefault(require("./behaviors"));

var _components = _interopRequireDefault(require("./components"));

var _Graphin = _interopRequireDefault(require("./Graphin"));

var _GraphinContext = _interopRequireDefault(require("./GraphinContext"));

var _registerGraphinForce = _interopRequireDefault(require("./layout/inner/registerGraphinForce"));

var _registerPresetLayout = _interopRequireDefault(require("./layout/inner/registerPresetLayout"));

var _shape = require("./shape");

var _utils = _interopRequireDefault(require("./utils"));

var _g = _interopRequireDefault(require("@antv/g6"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/** 注册 Graphin force 布局 */
(0, _registerGraphinForce.default)();
/** 注册 Graphin preset 布局 */

(0, _registerPresetLayout.default)();
/** 注册 Graphin Circle Node */

(0, _shape.registerGraphinCircle)();
/** 注册 Graphin line Edge */

(0, _shape.registerGraphinLine)();
/** 解构静态方法 */

var registerFontFamily = _Graphin.default.registerFontFamily;
/** export */

exports.registerFontFamily = registerFontFamily;
var _default = _Graphin.default;
exports.default = _default;
