// @antv/graphin/lib/components/ContextMenu/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _useContextMenu = _interopRequireDefault(require("./useContextMenu"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var defaultStyle = {
  width: '120px',
  boxShadow: '0 4px 12px rgb(0 0 0 / 15%)'
};

var container = /*#__PURE__*/_react.default.createRef();

var ContextMenu = function ContextMenu(props) {
  var bindType = props.bindType,
      children = props.children,
      style = props.style;
  var contextmenu = (0, _useContextMenu.default)({
    bindType: bindType,
    container: container
  });
  var visible = contextmenu.visible,
      x = contextmenu.x,
      y = contextmenu.y,
      item = contextmenu.item;
  var positionStyle = {
    position: 'absolute',
    left: x,
    top: y
  };

  if (typeof children !== 'function') {
    console.error('<ContextMenu /> children should be a function');
    return null;
  }

  var id = item && !item.destroyed && item.getModel && item.getModel().id || '';
  return /*#__PURE__*/_react.default.createElement("div", {
    ref: container,
    className: "graphin-components-contextmenu",
    style: Object.assign(Object.assign(Object.assign({}, defaultStyle), style), positionStyle),
    key: id
  }, visible && children(Object.assign(Object.assign({}, contextmenu), {
    id: id
  })));
};

var _default = ContextMenu;
exports.default = _default;
