// @antv/graphin/es/components/ContextMenu/index.js
import React from 'react';
import useContextMenu from './useContextMenu';
var defaultStyle = {
  width: '120px',
  boxShadow: '0 4px 12px rgb(0 0 0 / 15%)'
};
var container = /*#__PURE__*/React.createRef();

var ContextMenu = function ContextMenu(props) {
  var bindType = props.bindType,
      children = props.children,
      style = props.style;
  var contextmenu = useContextMenu({
    bindType: bindType,
    container: container
  });
  var visible = contextmenu.visible,
      x = contextmenu.x,
      y = contextmenu.y,
      item = contextmenu.item;
  var positionStyle = {
    position: 'absolute',
    left: x,
    top: y
  };

  if (typeof children !== 'function') {
    console.error('<ContextMenu /> children should be a function');
    return null;
  }

  var id = item && !item.destroyed && item.getModel && item.getModel().id || '';
  return /*#__PURE__*/React.createElement("div", {
    ref: container,
    className: "graphin-components-contextmenu",
    style: Object.assign(Object.assign(Object.assign({}, defaultStyle), style), positionStyle),
    key: id
  }, visible && children(Object.assign(Object.assign({}, contextmenu), {
    id: id
  })));
};

export default ContextMenu;
