// @antv/graphin/es/components/ContextMenu/useContextMenu.js
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

import React, { useEffect } from 'react';
import { GraphinContext } from '../../index';

var useContextMenu = function useContextMenu(props) {
  var _props$bindType = props.bindType,
      bindType = _props$bindType === void 0 ? 'node' : _props$bindType,
      container = props.container;

  var _React$useContext = React.useContext(GraphinContext),
      graph = _React$useContext.graph;

  var _React$useState = React.useState({
    visible: false,
    x: 0,
    y: 0,
    item: null,
    selectedItems: []
  }),
      _React$useState2 = _slicedToArray(_React$useState, 2),
      state = _React$useState2[0],
      setState = _React$useState2[1];

  var handleShow = function handleShow(e) {
    e.preventDefault();
    e.stopPropagation();
    var width = graph.get('width');
    var height = graph.get('height');

    if (!container.current) {
      return;
    }

    var bbox = container.current.getBoundingClientRect();
    var offsetX = graph.get('offsetX') || 0;
    var offsetY = graph.get('offsetY') || 0;
    var graphTop = graph.getContainer().offsetTop;
    var graphLeft = graph.getContainer().offsetLeft;
    var x = e.canvasX + graphLeft + offsetX;
    var y = e.canvasY + graphTop + offsetY; // when the menu is (part of) out of the canvas

    if (x + bbox.width > width) {
      x = e.canvasX - bbox.width - offsetX + graphLeft;
    }

    if (y + bbox.height > height) {
      y = e.canvasY - bbox.height - offsetY + graphTop;
    }

    if (bindType === 'node') {
      // 如果是节点，则x，y指定到节点的中心点
      // eslint-disable-next-line no-underscore-dangle
      var _ref = e.item && e.item.getModel(),
          PointX = _ref.x,
          PointY = _ref.y;

      var CenterCanvas = graph.getCanvasByPoint(PointX, PointY);
      var daltX = e.canvasX - CenterCanvas.x;
      var daltY = e.canvasY - CenterCanvas.y;
      x = x - daltX;
      y = y - daltY;
    }
    /** 设置变量 */


    setState(function (preState) {
      return Object.assign(Object.assign({}, preState), {
        visible: true,
        x: x,
        y: y,
        item: e.item
      });
    });
  };

  var handleClose = function handleClose() {
    setState(function (preState) {
      if (preState.visible) {
        return Object.assign(Object.assign({}, preState), {
          visible: false,
          x: 0,
          y: 0
        });
      }

      return preState;
    });
  };

  useEffect(function () {
    var handleSaveAllItem = function handleSaveAllItem(e) {
      setState(function (preState) {
        return Object.assign(Object.assign({}, preState), {
          selectedItems: e.selectedItems
        });
      });
    }; // @ts-ignore


    graph.on("".concat(bindType, ":contextmenu"), handleShow);
    graph.on('canvas:click', handleClose);
    graph.on('canvas:drag', handleClose);
    graph.on('wheelzoom', handleClose);

    if (bindType === 'canvas') {
      graph.on('nodeselectchange', handleSaveAllItem);
    }

    return function () {
      graph.off("".concat(bindType, ":contextmenu"), handleShow);
      graph.off('canvas:click', handleClose);
      graph.off('canvas:drag', handleClose);
      graph.off('wheelzoom', handleClose);
      graph.off('nodeselectchange', handleSaveAllItem);
    };
  }, [graph, bindType]);
  var x = state.x,
      y = state.y,
      visible = state.visible,
      item = state.item,
      selectedItems = state.selectedItems;
  return {
    oneShow: handleShow,
    onClose: handleClose,
    item: item,
    selectedItems: selectedItems,
    visible: visible,
    x: x,
    y: y
  };
};

export default useContextMenu;
