// @antv/graphin/lib/components/Legend/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _Node = _interopRequireDefault(require("./Node"));

var _useLegend2 = _interopRequireDefault(require("./useLegend"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var defaultStyle = {
  position: 'absolute',
  top: '0px',
  right: '0px'
}; // @ts-ignore

var Legend = function Legend(props) {
  var _props$bindType = props.bindType,
      bindType = _props$bindType === void 0 ? 'node' : _props$bindType,
      sortKey = props.sortKey,
      children = props.children,
      style = props.style;

  var _useLegend = (0, _useLegend2.default)({
    bindType: bindType,
    sortKey: sortKey
  }),
      dataMap = _useLegend.dataMap,
      options = _useLegend.options;

  return /*#__PURE__*/_react.default.createElement("div", {
    className: "graphin-components-legend",
    // @ts-ignore
    style: Object.assign(Object.assign({}, defaultStyle), style)
  }, children({
    bindType: bindType,
    sortKey: sortKey,
    // @ts-ignore
    dataMap: dataMap,
    // @ts-ignore
    options: options
  }));
};

Legend.Node = _Node.default;
var _default = Legend;
exports.default = _default;
