// @antv/graphin-components/lib/Combo/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _graphin = require("@antv/graphin");

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var __rest = void 0 && (void 0).__rest || function (s, e) {
  var t = {};

  for (var p in s) {
    if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0) t[p] = s[p];
  }

  if (s != null && typeof Object.getOwnPropertySymbols === "function") for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
    if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};

var defaultComboCfg = {
  members: [],
  type: 'rect',
  style: {
    fill: 'lightblue',
    stroke: 'blue',
    opacity: 0.2
  },
  padding: 10
};
/**
 * deep merge Combo config
 * @param defaultCfg
 * @param cfg
 */

var deepMergeCfg = function deepMergeCfg(defaultCfg, cfg) {
  var _defaultCfg$style = defaultCfg.style,
      DefaultCfg = _defaultCfg$style === void 0 ? {} : _defaultCfg$style,
      defaultOtherCfg = __rest(defaultCfg, ["style"]);

  var _cfg$style = cfg.style,
      style = _cfg$style === void 0 ? {} : _cfg$style,
      others = __rest(cfg, ["style"]);

  return Object.assign(Object.assign(Object.assign({}, defaultOtherCfg), others), {
    style: Object.assign(Object.assign({}, DefaultCfg), style)
  });
};

var Combo = function Combo(props) {
  var graphin = _react.default.useContext(_graphin.GraphinContext);

  var graph = graphin.graph;

  _react.default.useEffect(function () {
    var options = props.options;
    options.map(function (item) {
      return graph.createCombo( // @ts-ignore
      deepMergeCfg(defaultComboCfg, Object.assign({
        id: "".concat(Math.random())
      }, item)), item.members);
    });
    return function () {};
  }, [graph]);

  return _react.default.createElement(_react.default.Fragment, null);
};

var _default = Combo;
exports.default = _default;
