// @antv/graphin-components/lib/Legend/Node.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _graphin = require("@antv/graphin");

var _react = _interopRequireWildcard(require("react"));

require("./index.css");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var LegendNode = function LegendNode(props) {
  var _React$useContext = _react.default.useContext(_graphin.GraphinContext),
      legend = _React$useContext.legend,
      graph = _React$useContext.graph,
      theme = _React$useContext.theme; // 依然存在两个legend，graphin.context只是一个全局对象


  var _props$onChange = props.onChange,
      onChange = _props$onChange === void 0 ? function () {} : _props$onChange,
      style = props.style;
  var mode = theme.mode;
  var _legend$node = legend.node,
      defaultOptions = _legend$node.options,
      dataMap = _legend$node.dataMap;

  var _React$useState = _react.default.useState({
    options: defaultOptions
  }),
      _React$useState2 = _slicedToArray(_React$useState, 2),
      state = _React$useState2[0],
      setState = _React$useState2[1];
  /** 更新state依赖 */


  (0, _react.useEffect)(function () {
    setState({
      options: defaultOptions
    });
  }, [defaultOptions]);
  var options = state.options;

  var handleClick = function handleClick(option) {
    var checkedValue = Object.assign(Object.assign({}, option), {
      checked: !option.checked
    });
    var result = options.map(function (c) {
      var matched = c.value === option.value;
      return matched ? checkedValue : c;
    });
    setState({
      options: result
    });
    var nodes = dataMap.get(checkedValue.value);
    /** highlight */
    // const nodesId = nodes.map((c) => c.id);
    // apis.highlightNodeById(nodesId);

    nodes.forEach(function (node) {
      graph.setItemState(node.id, 'active', checkedValue.checked);
      graph.setItemState(node.id, 'inactive', !checkedValue.checked);
    });
    /** 给用户的回调函数 */

    onChange(checkedValue, result, props);
  };

  return _react.default.createElement("ul", {
    className: "graphin-components-legend-content",
    style: style
  }, options.map(function (option) {
    var label = option.label,
        checked = option.checked,
        color = option.color;
    var dotColors = {
      light: {
        active: color,
        inactive: '#ddd'
      },
      dark: {
        active: color,
        inactive: '#2f2f2f'
      }
    };
    var labelColor = {
      light: {
        active: '#000',
        inactive: '#ddd'
      },
      dark: {
        active: '#fff',
        inactive: '#2f2f2f'
      }
    };
    var status = checked ? 'active' : 'inactive';
    return _react.default.createElement("li", {
      key: option.value,
      className: "item",
      onClick: function onClick() {
        handleClick(option);
      },
      onKeyDown: function onKeyDown() {
        handleClick(option);
      }
    }, _react.default.createElement("span", {
      className: "dot",
      style: {
        background: dotColors[mode][status]
      }
    }), _react.default.createElement("span", {
      className: "label",
      style: {
        color: labelColor[mode][status]
      }
    }, label));
  }));
};

var _default = LegendNode;
exports.default = _default;
