// @antv/graphin/lib/layout/inner/tweak.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

/* eslint-disable no-param-reassign */
var getRandomPosition = function getRandomPosition() {
  return Math.round((Math.random() - 0.5) * 80);
};
/**
 *
 * @param currentData
 * @param options
 */


var width = 500;
var height = 300;

var tweak = function tweak(currentData, prevData) {
  var currNodes = currentData.nodes,
      currEdges = currentData.edges;
  var preNodes = prevData.nodes;
  /** 将图上之前节点的位置信息存储在positionMap中 */

  var positionMap = new Map();
  preNodes.forEach(function (item) {
    var id = item.id,
        x = item.x,
        y = item.y;
    positionMap.set(id, {
      x: x,
      y: y
    });
  });
  var incrementNodesMap = new Map();
  currNodes.forEach(function (node) {
    var id = node.id;

    var _positionMap$get = positionMap.get(id),
        x = _positionMap$get.x,
        y = _positionMap$get.y;

    if (!window.isNaN(x) && !window.isNaN(y)) {
      node.x = x;
      node.y = y;
    } else {
      incrementNodesMap.set(id, node);
    }
  });
  var incrementPositonMap = new Map();
  currEdges.forEach(function (edge) {
    var source = edge.source,
        target = edge.target;
    var nodeInSource = incrementNodesMap.get(source);
    var nodeInTarget = incrementNodesMap.get(target);
    var sourcePosition = positionMap.get(source);
    var positionInSource = !window.isNaN(sourcePosition.x) && !window.isNaN(sourcePosition.y);
    var targetPosition = positionMap.get(target);
    var positionInTarget = !window.isNaN(targetPosition.x) && !window.isNaN(targetPosition.y);

    if (nodeInSource && positionInTarget) {
      incrementPositonMap.set(source, {
        // ...nodeInSource,
        x: targetPosition.x + getRandomPosition(),
        y: targetPosition.y + getRandomPosition()
      });
    }

    if (nodeInTarget && positionInSource) {
      incrementPositonMap.set(target, {
        // ...nodeInTarget,
        x: sourcePosition.x + getRandomPosition(),
        y: sourcePosition.y + getRandomPosition()
      });
    }
  });
  currNodes.forEach(function (node) {
    var id = node.id;
    var position = positionMap.get(id);

    if (window.isNaN(position.x) || window.isNaN(position.y)) {
      position = incrementPositonMap.get(id) || {};
    }

    if (!window.isNaN(position.x) && !window.isNaN(position.y)) {
      node.x = position.x;
      node.y = position.y;
    } else {
      node.x = width / 2 + Math.round(Math.random() * width);
      node.y = height / 2 + Math.round(Math.random() * height);
    }
  });
  return Object.assign(Object.assign({}, currentData), {
    nodes: currNodes,
    edges: currEdges
  });
};

var _default = tweak;
exports.default = _default;
